modules/sdk-coin-btc/src/lib/keyPair.ts

```diff
@@ -0,0 +1,96 @@
+import { randomBytes } from 'crypto';
+import * as bip32 from 'bip32';
+import {
+  AddressFormat as BaseAddressFormat,
+  DefaultKeys,
+  isPrivateKey,
+  isPublicKey,
+  isSeed,
+  KeyPairOptions,
+  NotImplementedError,
+  Secp256k1ExtendedKeyPair,
+} from '@bitgo/sdk-core';
+import { address, crypto, networks } from 'bitcoinjs-lib';
+
+const DEFAULT_SEED_SIZE_BYTES = 16;
+
+// TODO: extend the AddressFormat enum in @bitgo/sdk-core with these values
+enum ExtendedAddressFormat {
+  mainnetBech32 = 'mainnetBech32',
+  testnetBech32 = 'testnetBech32',
+}
+
+export type AddressFormatType = BaseAddressFormat | ExtendedAddressFormat;
+export const AddressFormat = { ...BaseAddressFormat, ...ExtendedAddressFormat };
+
+/** Bitcoin key management. */
+export class KeyPair extends Secp256k1ExtendedKeyPair {
+  /**
+   * Public constructor. By default, creates a key pair with a random master seed.
+   *
+   * @param { KeyPairOptions } source Either a master seed, a private key (extended or raw), or a public key
+   *     (extended, compressed, or uncompressed)
+   */
+  constructor(source?: KeyPairOptions) {
+    super(source);
+    if (!source) {
+      const seed = randomBytes(DEFAULT_SEED_SIZE_BYTES);
+      this.hdNode = bip32.fromSeed(seed);
+    } else if (isSeed(source)) {
+      this.hdNode = bip32.fromSeed(source.seed);
+    } else if (isPrivateKey(source)) {
+      this.recordKeysFromPrivateKey(source.prv);
+    } else if (isPublicKey(source)) {
+      this.recordKeysFromPublicKey(source.pub);
+      // TODO: move this to Secp256k1ExtendedKeyPair
+      this.keyPair.compressed = source.pub.length === 66;
+    } else {
+      throw new Error('Invalid key pair options');
+    }
+
+    if (this.hdNode) {
+      this.keyPair = Secp256k1ExtendedKeyPair.toKeyPair(this.hdNode);
+    }
+  }
+
+  /**
+   * Bitcoin uncompressed public and private keys in hex format.
+   *
+   * @returns { DefaultKeys } The keys in the protocol default key format
+   */
+  getKeys(): DefaultKeys {
+    return {
+      pub: this.getPublicKey({ compressed: false }).toString('hex'),
+      prv: this.getPrivateKey()?.toString('hex'),
+    };
+  }
+
+  /**
+   * Get a bitcoin address in legacy or bech32 format.
+   *
+   * @param {AddressFormatType} format One of mainnet, testnet, mainnetBech32, or testnetBech32
+   * @returns {string} The address derived from the public key
+   */
+  getAddress(format?: AddressFormatType): string {
+    const compressed = this.keyPair.compressed;
+    const publicKeyHash160 = crypto.ripemd160(crypto.sha256(this.getPublicKey({ compressed })));
```

Comment:
why only support pay to public key hash addresses? These addresses are not compatible with bitgo multisig wallets.

Reply:
The `KeyPair` class is designed to handle a single set of keys. In order to manage multiple `KeyPair`s you would be better to add another abstraction layer like `MultiSigKeyChain` or `KeyStore` or something like that where you can add multisig-specific logic.

```
/** Manage a multisig account and their keys  */
export class MultiSigKeyChain {
  constructor(m: number, n: number) {
    ...
  }
  
  /**  Add a key pair with or without their private material. Private material is only recommended for testing  */
  addKeyPair(sigingKeyPair: KeyPair) { ... }

  /** Sign a transaction with all stored KeyPairs or a particular one if pub is passed as a parameter. */
  sign(tx: string, pub?: PublicKey) {}

  /** Get a multisig compatible address. */
  getAddress(format: AddressFormat) { .. }
}
```

This is what's been done for Algorand (in a non-generic way).

Adding the address format and throwing an error is probably a good idea here since a multisig address cannot be generated under the 1 key primitive. I leave the multisig class implementation to you guys unless you want me to give it a stub in a follow-up PR.

Reply:
You actually already have that multisig management here: https://github.com/BitGo/BitGoJS/blob/549fbbcded358e0207da298e0278dfdb09a4d787/modules/utxo-lib/src/bitgo/wallet/WalletKeys.ts#L73

And that's where it belongs since it is BitGo specific, so scratch the `MultiSigKeyChain` suggestion. This class is designed to deal with a single `KeyPair`.